### ISSUE TITLE
Not obvious how to get text editor / document from editor view content

### ISSUE THREAD

**Issue by mhutch** [mono/monodevelop#3559]
Now that `SourceEditorView` and `TextEditorData` are internal, this makes it more difficult to do things with a `SourceEditorView` instance.

For example, when implementing `IToolboxDynamicProvider.GetDynamicItems`, the `SourceEditorView` is passed in as the `IToolboxConsumer` parameter. As far as I can tell, there is no way from this to get the `Document`, the `TextEditor or the `ITextDocument`.

Previously I would have cast it to a `ViewContent` then used `GetContent<TextEditorData>` to get the editor data structures, or cast it to a `SourceEditorView` and accessed the `Document` property.

Now to get an editor I have to cast to `ViewContent` then access `WorkbenchWindow.Document.TextEditor`

Perhaps `ViewContent` should have a `Document` property? Perhaps `IToolboxDynamicProvider.GetDynamicItems` should be passed a `Document`? Perhaps `ViewContent.GetContent` should allow getting the TextEditor?

> VS bug #577080

**Comment by mkrueger** [comment]
GetContent<IToolboxDynamicProvider> () should work

**Comment by mhutch** [comment]
`ViewContent.WorkbenchWindow.Document.GetContent<T>();` on a `SourceEditorView` returns the TextEditor. `ViewContent.GetContent<TextEditor>()` returns null. This is pretty confusing.

**Comment by mkrueger** [comment]
SourceEditorView shouldn't be used directly.

TextEditorViewContent is what the text editor contains and GetContent<TextEditor> () works there.

**Comment by mkrueger** [comment]
Note to the tech background: GetContent can only work no lower layers and SourceEditorView doesn't contain a TextEditor anymore - TextEditorViewContent does.

SEV used to do it but with the infrastructure change TextEditor moved to IDE.

**Comment by mhutch** [comment]
I'm not trying to use it directly. An instance of it is being passed to me so I have no choice.

**Comment by mkrueger** [comment]
And that's the error - you shouldn't get any instance.

**Comment by mhutch** [comment]
Then the bug is valid 🙂

**Comment by mkrueger** [comment]
GetDynamicItems doesn't get any view content and IToolboxConsumer doesn't have any connection to a view content. Upcasts are always problematic and can have side effects. How about changing the interface of that ?

We don't want gtk implementations on the ide text editor layer so it's not easy to move the implementation down.

**Comment by mhutch** [comment]
The problem here is that the closing down of the editor API has basically broken the IToolboxDynamicProvider API.

The IToolboxDynamicProvider was always intended to be able to upcast the IToolboxConsumer parameter to a real view (e.g. a designer) in order to get information from it that could be used to determine which items to show in the toolbox. Unfortunately, now that the editor API is internal, when the IToolboxConsumer in an editor, it cannot be directly upcasted to anything useful.

For example - I want to show text items when the view is a text editor with a XAML file. Determining that the view is a text editor now involves reflecting internal code, and fetching its mimetype is now VERY cumbersome.

Previously, this code would have looked like this, which is pretty straightforward:

```csharp
public IEnumerable<ItemToolboxNode> GetDynamicItems(IToolboxConsumer consumer)
{
    if (consumer is SourceEditorView editor && editor.MimeType=="application/xaml") {
        return GetItems(editor);
    }
    return null;
}
```

Now it goes through about 4 levels of indirection, which is not obvious at all:

```csharp
public IEnumerable<ItemToolboxNode> GetDynamicItems(IToolboxConsumer consumer)
{
    TextEditor editor = null;
    if (consumer is ViewContent vc && consumer is ITextFile) {
        editor = vc.WorkbenchWindow.Document.GetContent<TextEditor>();
    }
    if (editor != null && editor.MimeType=="application/xaml") {
        return GetItems(editor);
    }
    return null;
}
```

**Comment by mhutch** [comment]
cc @slluis

**Comment by mkrueger** [comment]
You can't rely on upcasting to a special class - that's just a broken API design. It's not obvious for the implementor of that interface what's required here. 

I would suggest adding a GetViewContent method to that inferface - this can give the correct view content. So designing a non obvious API gives non obvious results.

I can't move all the GTK stuff to the core TextView just because of 3 lines indirections in the toolbox addin.

**Comment by mhutch** [comment]
What do you mean, a GetViewContent method? What will it do other than just casting the consumer to a ViewContent? There will won't be a meaningful text editor viewcontent that I can access.

If you mean GetContent, that's problematic too. If it operates on the Document.GetViewContent, then GetContent<TextEditor> will return an editor _even if the focused viewcontent is a designer subview_. If it operates on the ViewContent, then it'll return null, as _GetContent<TextEditor> on the editor viewcontent returns null_.

Passing a Document to the IToolboxDynamicProvider would definitely help, but we still need some way to determine that the consumer is the source editor. Maybe checking for `DefaultItemDomain == "Text"` is the best way, but it's not obvious.

Maybe we should add something like this?

```csharp
public static class ToolboxExtensions
{
    public static T GetContent<T> (this IToolboxConsumer consumer) => (consumer as ViewContent)?.GetContent<T> ();
    public static Document GetDocument (this IToolboxConsumer consumer) => (consumer as ViewContent)?.WorkbenchWindow?.Document;
    public static bool IsTextEditor () (this IToolboxConsumer consumer, out TextEditor editor)
    {
        editor = consumer.DefaultItemDomain == "Text"? GetDocument(consumer)?.TextEditor : null;
        return editor != null;
    }
}
```

Then I could write

```csharp
public IEnumerable<ItemToolboxNode> GetDynamicItems(IToolboxConsumer consumer)
{
    if (consumer.IsTextEditor(out TextEditor editor) && editor.MimeType=="application/xaml") {
        return GetItems(editor);
    }
    return null;
}
```

I still think ViewContent.GetContent<TextEditor>() should work on the source editor viewcontent.

**Comment by mkrueger** [comment]
The source editor view content is no view content anymore - it's only one for historical reasons. The text editor view content moved to the ide. The source editor view content is not added as view content directly so it doesn't work as view content.

**Comment by mhutch** [comment]
Maybe the IDE level one should be the consumer then

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.6.0.377 (master)

Author: Mikayla Hutchinson
Commit: 2ea2440fd15cfcb62859c6958defed327e5d8f22 (mono/monodevelop)

**Comment by xamarin-release-manager** [comment]
Fixed in version 7.6.0.377 (master)

Author: Mikayla Hutchinson
Commit: 2ea2440fd15cfcb62859c6958defed327e5d8f22 (mono/monodevelop)

**Comment by mhutch** [comment]
I added workarounds to improve the API, but the issue with the editor passing a basically unusable object to the toolbox provider still stands.

**Comment by mkrueger** [comment]
It's not passing a unusuable object - it's passing an IToolboxDynamicProvider object.
You're assuming that it can be upcasted to a specific object of your choice - which is not part of any API design I ever encountered. 

Just add a GetViewContent() method to the IToolboxDynamicProvider and I can return a valid view content. Relying on upcasts is always bad. Maybe that interface might be implemented by no view content at all ?

**Comment by mhutch** [comment]
It's passing a consumer, not a provider.

We can't change interfaces because it breaks binary compat.

The intent of the API was that clients could conditionally upcast the consumer to determine whether it was a specific consumer, and if so, they could use the upcasted consumer.

Due to the editor API changes it's no longer possible to do that, so it's no longer possible for a provider to cleanly interact with the editor.

Maybe that means we need to change the API, but ignoring the use case isn't the answer.

**Comment by mkrueger** [comment]
The answer is that this change is not going to change - it's a wrong design.
Upcasting is nothing that an API should enforce to do or rely on a specific class the object can be upcasted to.

So it's not ignoring - the answer is : we can not change it. (Without breaking the API)

**Comment by mkrueger** [comment]
@Therzok / @DavidKarlas what's your pov on that issue ?

**Comment by Therzok** [comment]
cc @KirillOsenkov @DavidKarlas given we've added ITextBuffer/ITextView support, should this be closed?

**Comment by mhutch** [comment]
Yes `GetContent<ITextView>()` does not have the issue that `GetContext<TextEditor>()` had.

I still think the `IToolboxDynamicProvider` interface has design issues as upcasting is generally a bad pattern, but I [papered over that with extension methods](https://github.com/mono/monodevelop/blob/305bddc1e9dcd90caf25f8bed1a042efc317528b/main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Toolbox/ToolboxExtensions.cs)

**Comment by xamarin-release-manager** [comment]
Fixed in version 8.6.0.802 (master)

Author: Mikayla Hutchinson
Commit: 2ea2440fd15cfcb62859c6958defed327e5d8f22 (xamarin/monodevelop)
